Page directive registration

DanteDeRuwe opened the pull request:
This is related to #4.

This change makes sure one could write
```ts
export function setup(api) {
  api.defineBlazorReferences([...], {
    includePages: true, //set to false or omit to not auto discover
  });
}
```

All components marked with a `@page` directive will then be registered. They will also pick up route parameters through `args.match.params` if possible.

Their names are sanitized: Any leading slashes are removed and everything that is not alphanumeric, an underscore, or a dash gets replaced with an underscore. This is important to know for the creation of a codegen that will register all names on compile time, later.

DanteDeRuwe committed:
Merge branch 'develop' into feature/references

DanteDeRuwe committed:
Implement basic @page directive registration

using LoadComponentFromLibrary args ( i.e. { includePages: true }). 
Related to #4

DanteDeRuwe committed:
Parse boolean option correctly + add logging and exception handling

DanteDeRuwe committed:
Refactor extension methods

DanteDeRuwe committed:
Implement better string support

DanteDeRuwe committed:
Fix loading non-pages

DanteDeRuwe committed:
Merge remote-tracking branch 'origin/develop' into feature/references

DanteDeRuwe committed:
Fix after merge

DanteDeRuwe committed:
Merge remote-tracking branch 'origin/develop' into feature/references

FlorianRappl commented on src/Piral.Blazor.Core/ComponentActivationService.cs:
```diff
@@ -80,22 +88,80 @@ public void DeactivateComponent(string componentName)
         }
 
-        public void LoadComponentsFromAssembly(Assembly assembly)
+        public void LoadComponentsFromAssembly(Assembly assembly, IDictionary<string, object> args = default)
```
Let's remove these and solve it as follows:

1. A codegen to bring in the `registerPage` calls by reading out the assembly's meta (being picky on the page attributes) and then bringing them in via a conversion of their route (blazor format to react router format) and calling `fromBlazor` with `page-${original-page-route}` as name.
2. The component resolver in loadcomponents does not only collect expose pilet attributes, but also page attributes registered as `page-${original-page-route}`

FlorianRappl commented:
The original mechanism is still:

```ts
export function setup(api) {
  api.defineBlazorReferences([...]);
}
```

To register pages:

```ts
import { registerDependencies, registerBlazorPages } from './blazor.codegen';

export function setup(api) {
  registerDependencies(api); // calls api.defineBlazorReferences with all DLLs
  registerBlazorPages(api); // calls api.registerPage per page
  // ...
}
```

DanteDeRuwe committed:
Always register pages (with "page-" prefix)

It's not necessary, and it makes it  so that piral-blazor does not have to be changed. The pages will now also be registered with a "page-" prefix.

DanteDeRuwe commented on src/Piral.Blazor.Core/ComponentActivationService.cs:
```diff
@@ -124,44 +120,26 @@ private Type GetComponent(string componentName)
+                    string template = member.GetCustomAttribute<RouteAttribute>(false)?.Template;
+                    return template is null
+                        ? null
+                        : $"page-{Sanitize(template)}"; //pages have a "page-" prefix and get sanitized
```
@FlorianRappl  I did not go with `page-${original-page-route}`, but sanitized the route first. This is because piral-blazor has this line:

https://github.com/smapiot/piral/blob/89b2928224f69d0cc534f02f3a9d2d8ffb5d9245/src/converters/piral-blazor/src/converter.ts#L38

Which fails if the id given contains characters that are not allowed in css selectors (like `/` etc)

This can be resolved by using `getElementById`, or only somehow only sanitizing the HTML id's. But for now, I sanitize the component name itself. I use the same regex in the codegen too.

Would like your thoughts on this.

FlorianRappl replied:
That's great ahead thinking. Yeah the sanitization makes sense. We only need to be careful to apply it consistently.

FlorianRappl commented on src/Piral.Blazor.Core/ComponentActivationService.cs:
```diff
@@ -83,19 +91,55 @@ public void DeactivateComponent(string componentName)
+            switch (attributeType)
+            {
+                case Type _ when attributeType == typeof(RouteAttribute):
+                    string template = member.GetCustomAttribute<RouteAttribute>(false)?.Template;
```
var

FlorianRappl commented on src/Piral.Blazor.Core/ComponentActivationService.cs:
```diff
@@ -83,19 +91,55 @@ public void DeactivateComponent(string componentName)
+        private static string Sanitize(string value)
+        {
+            string val = value.StartsWith("/") ? value.Substring(1) : value;
+            return Regex.Replace(val, @"[^\w\-]", "_");
```
Maybe have it `+` such that we don't see `___` etc. but only `_`.

FlorianRappl commented on src/Piral.Blazor.Core/ComponentActivationService.cs:
```diff
@@ -83,19 +91,55 @@ public void DeactivateComponent(string componentName)
+        /// </summary>
+        private static string Sanitize(string value)
+        {
+            string val = value.StartsWith("/") ? value.Substring(1) : value;
```
var

FlorianRappl commented on src/Piral.Blazor.Core/ComponentActivationService.cs:
```diff
@@ -83,19 +91,55 @@ public void DeactivateComponent(string componentName)
+                var toRegister = GetComponentNamesToRegister(componentType, AttributeTypes);
+                foreach (string componentName in toRegister)
+                {
+                    if (componentName is null) continue;
```
Can this be a real case? I think the `GetComponentNamesToRegister` should only return valid names.

Also maybe call it `componentNames` instead of `toRegister`. The latter is implied by the function.

FlorianRappl commented on src/Piral.Blazor.Core/Extensions.cs:
```diff
@@ -22,50 +38,57 @@ static class Extensions
+            try
+            {
+                var routeParams = JsonSerializer.Deserialize<Match>(JsonSerializer.Serialize(args["match"]))?.@params;
+                if(!routeParams.IsNullOrEmpty()) allArgs.Add(routeParams);
```
Style (space after `if` please) 
Introduce curly brackets (*always*  use block statements)

FlorianRappl commented on src/Piral.Blazor.Core/Extensions.cs:
```diff
@@ -22,50 +38,57 @@ static class Extensions
-
-            if (value == null)
+            
+            if (value == null) return propType.GetDefaultValue();
```
Block statements (see comment above).

FlorianRappl commented on src/Piral.Blazor.Core/Extensions.cs:
```diff
@@ -22,50 +38,57 @@ static class Extensions
+        
         public static object NormalizeValue(this Type type, string key, object value)
         {
             var property = type.GetProperty(key);
```
I am not sure about the changes in this function. Why should it change? Why so many changes?

DanteDeRuwe replied:
You are right. This was due to the previous implementation where I had args passed in from js. Then, they would arrive as strings and needed to convert them to bools (bool.Parse...). Will revert!

DanteDeRuwe replied:
I just noticed that this is also necessary for the route params to function properly. If the parameter in Blazor needs to be an integer, the route param that is found in `args.match.params` will still be a string. So it needs to be parsed correctly.

FlorianRappl replied:
Alright, but let's split it up a bit to have the dedicated conversion in this function and have a target-specified conversion in another function. The target-specified conversion needs a string to start with (it only supports string->X where X may be one of Int32, Int64, Double, Float, Guid, ...).

DanteDeRuwe replied:
Solving this with `TypeDescriptor.GetConverter` as discussed

FlorianRappl commented on src/Piral.Blazor.Core/Program.cs:
```diff
@@ -31,6 +32,7 @@ public static async Task Main(string[] args)
         private static void Configure(WebAssemblyHost host)
         {
             JSBridge.Configure(host.Services.GetRequiredService<HttpClient>());
+            Extensions.Configure(host.Services.GetService<ILoggerFactory>());
```
This is an anti-pattern. The only reason to do it with the bridge is that the bridge is already an anti pattern and this was the only way.

For the above you have better options. If you *really* want a logger in there (which you shouldn't) then bring it in as an argument. Then its up to the caller. In general, however, it should be the caller of an extension to do the logging. Extensions are low-level functions that don't do the logging (or error handling). They define the working parts.

DanteDeRuwe committed:
Change sanitization

DanteDeRuwe committed:
Revert to old NormalizeValue function

DanteDeRuwe committed:
Remove logger and unused method from Extensions

DanteDeRuwe committed:
Refactor code

for more consistent code style

DanteDeRuwe committed:
Try to convert every parameter using a converter

DanteDeRuwe committed:
Add Piral.Blazor.Analyzer to analyze pages

DanteDeRuwe committed:
Refactor analyzer

FlorianRappl reviewed (approved):
LGTM!
